## Data-driven optimization and decision making - Assignment 3
Juha Reinikainen

Solve any benchmark problems (K=2 and 5, n=10) with ParEGO and LCB. Start with 109
design points. Compare the hypervolume of the solutions after 100 exact function
evaluations.


In [8]:
import numpy as np
from sklearn.gaussian_process import GaussianProcessRegressor
from desdeo_problem.testproblems.TestProblems import test_problem_builder
from desdeo_tools.utilities.quality_indicator import hypervolume_indicator
from pymoo.factory import get_problem

from parego import ParEGO
from lcb import optimize_with_lcb
from sampling import create_data


In [9]:
seed = 1
max_evals = 100

dtlz2_pymoo = get_problem("dtlz2", n_var=10, n_obj=2)
dtlz5_pymoo = get_problem("dtlz5", n_var=10, n_obj=5)
dtlz2 = test_problem_builder("DTLZ2", n_of_objectives=2, n_of_variables=10)
dtlz5 = test_problem_builder("DTLZ5", n_of_objectives=5, n_of_variables=10)


In [10]:
#DTLZ2

X, y = create_data(dtlz2, 109, seed)
surrogates = [GaussianProcessRegressor() for _ in range(2)]
individuals, solutions = optimize_with_lcb(dtlz2, surrogates, X, y, max_evals)
nadir_dtlz2 = np.max(solutions, axis=0)
hv_lcb = hypervolume_indicator(solutions, nadir_dtlz2)
print("dtlz2 lcb hv", hv_lcb)

parego = ParEGO()
X, y = create_data(dtlz2_pymoo, 109, seed)
individuals, solutions = \
    parego.optimize(dtlz2_pymoo, GaussianProcessRegressor(),
                    X, y, max_evals, 10, seed)
hv_parego = hypervolume_indicator(solutions, nadir_dtlz2)
print("dtlz2 parego hv", hv_parego)


dtlz2 lcb hv 3.1150908270394955
dtlz2 parego hv 1.647657050769505


In [11]:
#DTLZ5

X, y = create_data(dtlz5, 109, seed)
surrogates = [GaussianProcessRegressor() for _ in range(5)]
individuals, solutions = optimize_with_lcb(dtlz5, surrogates, X, y, max_evals)
nadir_dtlz5 = np.max(solutions, axis=0)
hv_lcb = hypervolume_indicator(solutions, nadir_dtlz5)
print("dtlz5 lcb hv", hv_lcb)

parego = ParEGO()
X, y = create_data(dtlz5_pymoo, 109, seed)
individuals, solutions = \
    parego.optimize(dtlz5_pymoo, GaussianProcessRegressor(),
                    X, y, max_evals, 10, seed)
hv_parego = hypervolume_indicator(solutions, nadir_dtlz5)
print("dtlz5 parego hv", hv_parego)

dtlz5 lcb hv 0.5523348967202094
dtlz5 parego hv 0.06240745908861826


| Instance             | HV (ParEGO) | HV (LCB) |
|  --                  | --          | --       |
| DTLZ2 (K = 2, n = 10) |  1.65           |   3.12       |
| DTLZ5 (K = 5, n = 10) |  0.062          |     0.55     |

LCB performs significantly better on both DLTZ2 and DTLZ5